app/services/allocations/updater.rb

```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+module Allocations
+  class Updater
+    include Eventable
+
+    attr_accessor :allocation_params, :allocation_id, :allocation, :created_by, :doorkeeper_application_owner
+
+    def initialize(allocation_params:, allocation_id:, doorkeeper_application_owner:, created_by:)
+      self.allocation_params = allocation_params
+      self.allocation_id = allocation_id
+      self.doorkeeper_application_owner = doorkeeper_application_owner
+      self.created_by = created_by
+    end
+
+    def call
+      self.allocation = Allocation.find(allocation_id)
+      existing_date = allocation.date
+
+      allocation.assign_attributes(allocation_params[:attributes])
+      allocation.validate!
+
+      update_move_dates if allocation.date != existing_date
+
+      allocation.save!
+    end
+
+  private
+
+    def update_move_dates
+      allocation.transaction do
+        allocation.moves.each do |move|
+          move.date = allocation.date
+          move.save!
+
+          create_automatic_event!(
+            eventable: move,
+            event_class: GenericEvent::MoveDateChanged,
+            details: { date: move.date.iso8601 },
+          )
+
+          Notifier.prepare_notifications(topic: move, action_name: 'update_status')
```

Comment:
this should be 'update', as the status doesn't change

Reply:
since this is called before `allocation.save!`, if the save fails, the notifications will still be sent, maybe not a problem but something to keep in mind

Reply:
Good spot on both counts. With the `update_status` I thought I had copied what the move controller does but I had misread it. And the test condition I was using to catch the allocation failing to save was also causing the move to fail to save, so I've improved the tests to catch the specific scenario you outlined and moved the `allocation.save!` inside the transaction so that it won't send the notifications (as they won't be created in the DB).